app/api/orders.py

```diff
@@ -74,6 +74,10 @@ def before_create_object(self, data, view_kwargs):
         :param view_kwargs:
         :return:
         """
+        if data.get('status') == 'cancelled' or data.get('status') == 'expired':
```

Comment:
OK, what if I set it to placed? Or hoola_boola?

Reply:
@iamareebjamal got your point.
I've a doubt, whenever any order is created on FE it is created with `initializing` status and when the user fills up the order form and hit on `Proceed` button a PATCH request is sent to update the status to either - `completed`, `placed` or `pending` and if user fails to fill it up in 15 minutes, it is changed to `expired`. 
So I just wanted to know should I only allow order with `initializing` status in `before_create_object`. Or should I allow status from - `'completed', 'initializing', 'pending', 'placed'`?

Reply:
Ask yourself, can the client set an order to be completed or placed before it is actually placed or completed. If so, how can I prevent that?

Then ask, can a client fake an order to be filled and just send the PATCH request before actually filling required details. If so, how do I prevent that?

In the last step, ask yourself. Can a client set an expired/pending/completed/placed order to be initializing state or any previous state. If so, how do I prevent that?

In short you have to ensure that an order follows a model of [Deterministic Finite State Automata](https://en.wikipedia.org/wiki/Deterministic_finite_automaton)

Order should have a starting state and final state and can only go from one state to another state with a precondition and not change after it is in final state

Reply:
I'm planning to implement this -
 - Allow only `initializing` status in `before_create_object` (It'll also block any creation of order object by making direct API call) (POST request)
 - Valid status transitions (PATCH request) :
```
1. initializing  ->  completed, expired, placed, pending
2. completed  ->  cancelled (only non-paid orders)
3. expired  ->  None
4. placed  ->  cancelled 
5. pending  ->  completed, expired
6. cancelled  ->  None
```

Reply:
Even then the transitions should be validated

Reply:
@iamareebjamal please review the checks applied on the transitions